### ISSUE TITLE
求助求助，llama模型生成第一个token时，有3个代码块耗时严重

### ISSUE THREAD

**Issue by AnShengqiang** [ztxz16/fastllm#350]
**前言**
感谢开发者，研发出了如此易理解、好部署、配件完善的加速库🎉🎉🎉，真的很棒，对我很有帮助😊😊😊！！！

**问题描述**
有大量业务场景，仅需要模型生成单个token，如：新闻分类、逻辑推断、情感分析、关系提取、语种检测...。
在此类场景下，fastllm库中的llama模型实现（其它模型可能也存在）存在一个严重问题：随着batch size增大，耗时线性增长😰。
这个问题其他用户也复现了，见issue：[ISSUE 337](https://github.com/ztxz16/fastllm/issues/337)

**复现细节**
- 硬件：显卡4090，内存cpu管够。
- 模型：LlamaModel
- 接口：batch_response

```
import pyfastllm
model_path = "tokenizer path"
tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code = True)

flm_model = pyfastllm.LlamaModel()
flm_model.load_weights("xxx_fp16.flm")

prompt = """..."""  # around 300 words

config = pyfastllm.GenerationConfig()
config.max_length = 1

batch_size = 16 # batch size=16时间是8的2倍（其它数值均尝试过，基本呈线性增长）
r = flm_model.batch_response([prompt]*batch_size, None, config)
```

**问题分析**
经过大量的耗时分析工作（调试、日志打印、分析工具等），我将耗时定位到了3个地方。
文件：llama.cpp
方法：std::vector <int> LlamaModel::ForwardBatch(int batch, const fastllm::Data &inputIds, const fastllm::Data &attentionMask,
                            const fastllm::Data &positionIds, std::vector<std::pair<Data, Data>> &pastKeyValues,
                            const GenerationConfig &generationConfig, const LastTokensManager &lastTokens,
                            std::vector <std::vector <float>*> *retLogits)
耗时点1：
PermuteSelf(q, {0, 2, 1, 3});
PermuteSelf(k, {0, 2, 1, 3});
PermuteSelf(v, {0, 2, 1, 3});

耗时点2：
CatDirect(pastKey, k, 1);
CatDirect(pastValue, v, 1);

耗时点3：
logits.ToDevice(DataDevice::CPU);

这几点不起眼的操作，在第一个token输出时，耗时**远大于**😨真正的前向计算操作。

**最后**
由于我才疏学浅😞，难以解决此问题，因此请求帮助，希望开发者大大们协助解决此问题，谢谢♥

希望这个项目越来越好，star✨过万！

**Comment by siemonchan** [comment]
@AnShengqiang llama.cpp代码比较老，里面算完logits后直接传回CPU了，可以参考chatglm.cpp里改一下，先topk再回CPU

**Comment by AnShengqiang** [comment]
![G(1K(E%UP0JZ8OZSTINBU$3](https://github.com/ztxz16/fastllm/assets/12854979/abc70135-c4b6-4b52-8a61-9e99f20d678d)

依次打印了各部分耗时:
总共1.51s耗时：3个操作加起来耗费 1.34s（PermuteSelf耗费0.431s，CatDirect耗费0.457s，ToDevice耗费0.455s）🫡

**Comment by AnShengqiang** [comment]
> @AnShengqiang llama.cpp代码比较老，里面算完logits后直接传回CPU了，可以参考chatglm.cpp里改一下，先topk再回CPU

这个可以的，前两个操作有希望规避不。这两个操作，也是随着batch size增大，耗时线性增长，能规避最好了。

**Comment by wildkid1024** [comment]
你是怎么统计时间的？cuda 同步之前测试的时间不对，使用nv专用的性能测试工具。

**Comment by AnShengqiang** [comment]
最新进展： #351 已经解决了后两个耗时的问题。速度提升巨大🎉🎉🎉感谢 @ztxz16 与 @siemonchan 的贡献 🫡

目前仅剩余第一个耗时问题。

**Comment by AnShengqiang** [comment]
llama.cpp的ForwardBatch方法中
PermuteSelf(q, {0, 2, 1, 3});
PermuteSelf(k, {0, 2, 1, 3});
PermuteSelf(v, {0, 2, 1, 3});
耗时依然很严重，占比很高（在首个token生成时），能否继续想办法优化或规避？ @ztxz16 @siemonchan 🙏🙏🙏

**Comment by yiguanxian** [comment]
你这个耗时问题是稳定复现的？
